samza-api/src/main/java/org/apache/samza/util/RateLimiter.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.util;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.samza.annotation.InterfaceStability;
+import org.apache.samza.config.Config;
+import org.apache.samza.task.TaskContext;
+
+/**
+ * A rate limiter interface used by Samza components to limit throughput of operations
+ * against a resource. Operations against a resource are represented by credits.
+ * Resources could be streams, databases, web services, etc.
+ *
+ * <p>
+ * This interface supports two categories of policies: tagged and non-tagged.
+ * Tagged rate limiter is used, when further differentiation is required within a resource.
+ * For example: messages in a stream may be treated differently depending on the
+ * overall situation of processing; or read/write operations to a database.
+ * Tagging is the mechanism to allow this differentiation.
+ *
+ * <p>
+ * The following types of invocations are provided
+ * <ul>
+ *   <li>Block indefinitely until requested credits become available</li>
+ *   <li>Block for a provided amount of time, then return available credits</li>
+ *   <li>Non-blocking, returns immediately available credits</li>
+ * </ul>
+ *
+ */
+@InterfaceStability.Unstable
+public interface RateLimiter extends Serializable {
+
+  /**
+   * Initialize this rate limiter, this method should be called during container initialization.
+   *
+   * @param config job configuration
+   * @param taskContext task context that owns this rate limiter
+   */
+  void init(Config config, TaskContext taskContext);
```

Comment:
It will be nicer if we don't do this init() but directly create the rate limiter inside init by passing in the task count. Then you can have a static factory method which does this thing: 

RateLimiter limiter = RateLimiter.create(config, taskContext);

Having init() is a little bit out of place since it's not a full life cycle support (no close(), start(), stop() ...). No strong opinion for this, just some thoughts.